pkg/controller/controller.go

```diff
@@ -532,6 +532,26 @@ func (p *csiProvisioner) Provision(options controller.ProvisionOptions) (*v1.Per
 		return nil, capErr
 	}
 
+	if options.PVC.Spec.DataSource != nil && (rc.clone || rc.snapshot) {
+		contentSource := rep.GetVolume().ContentSource
+		if contentSource == nil {
+			sourceErr := fmt.Errorf("volume content source missing")
+			delReq := &csi.DeleteVolumeRequest{
+				VolumeId: rep.GetVolume().GetVolumeId(),
+			}
+			delReq.Secrets = provisionerCredentials
+			ctx, cancel := context.WithTimeout(context.Background(), p.timeout)
+			defer cancel()
+			_, err := p.csiClient.DeleteVolume(ctx, delReq)
+			if err != nil {
+				sourceErr = fmt.Errorf("%v. cleanup of volume %s failed, volume is orphaned: %v", sourceErr, pvName, err)
+			}
+			return nil, sourceErr
+		}
+
+		klog.Infof("created volume %s from snapshot %s successfully", pvName, contentSource.GetSnapshot().SnapshotId)
```

Comment:
This can be clone too, not just snapshot right?

Reply:
Yes, it can be clone too.